File: project_euler/Problem024/Problem024.py
from typing import List

class Problem024:
    def solution(self) -> None:
        target, elements = 1000000, [0, 1, 2, 3, 4, 5, 6, 7, 8, 9]
        self.__permutation(target - 1, elements, len(elements), 0)

        result = 0
        for element in elements:
            result = result * 10 + element
        print(result)

    def __permutation(self, sequence: int, elements: List[int], length: int, start: int) -> None:
        if 1 == length - start:
            return

        count = self.__get_factorial(length - start - 1)
        index = sequence // count + start
        if index != start:
            auxiliary = elements[index]
            elements[start + 1:index + 1] = elements[start:index]
            elements[start] = auxiliary
        start += 1
        self.__permutation(sequence % count, elements, length, start)

    def __get_factorial(self, value: int) -> int:
        result = 1
        for i in range(2, value + 1):
            result *= i
        return result

File: project_euler/Problem024/Problem024.go
package project_euler

import (
    "fmt"
)

func Problem024() {
    target := int64(1000000)
    elements := []int{0, 1, 2, 3, 4, 5, 6, 7, 8, 9}
    permutation(target - 1, elements, len(elements), 0)

    result := int64(0)
    for _, element := range elements {
        result = result * 10 + int64(element)
    }
    fmt.Println(result)
}

func permutation(sequence int64, elements []int, length int, start int) {
    if 1 == length - start {
        return
    }

    count := getFactorial(length - start - 1)
    index := int(sequence / count) + start
    if index != start {
        auxiliary := elements[index]
        for i := index; i > start; i-- {
            elements[i] = elements[i - 1]
        }
        elements[start] = auxiliary
    }
    start += 1
    permutation(sequence % count, elements, length, start)
}

func getFactorial(value int) int64 {
    result := int64(1)
    for i := 2; i <= value; i++ {
        result *= int64(i)
    }
    return result
}
